### PR TITLE
RHOL-306: Support for bundle

### PR DESCRIPTION
**Pull request by goral09**
## Overview
Handle `bundle` operator from rholang.

### Does this PR relate to an RChain JIRA issue? 
[JIRA ticket.](https://rchain.atlassian.net/browse/RHOL-306)

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
None.

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 1]
```diff
@@ -526,6 +530,15 @@ object ProcNormalizeMatcher {
         ProcVisitOutputs(input.par.prepend(foldedNew), bodyResult.knownFree)
       }
 
+      case b: PBundle =>
+        val targetResult = normalizeMatch(b.proc_, input.copy(par = VectorPar()))
+        if (targetResult.par.wildcard || targetResult.par.freeCount > 0) {
+          throw UnexpectedBundleContent(
+            s"Bundle's content shouldn't have free variables or wildcards.")
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Style issue: we shouldn't use s-strings unless we're actually substituting something in it.
knownFree contains all the line, column numbers where the free variables occurred.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala [hunk 2]
```diff
@@ -728,6 +730,34 @@ class ProcMatcherSpec extends FlatSpec with Matchers {
...
+        .withLocallyFree(BitSet(0))
+
+    result.par should be(expectedResult)
+    result.knownFree should be(inputs.knownFree)
+  }
+
+  /** Example:
+    * bundle { _ | x }
+    */
+  it should "not count free variable and the wildcard from the body" in {
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: The title of this test should change.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala [hunk 3]
```diff
@@ -188,3 +188,38 @@ class EvalSubSpec extends FlatSpec with Matchers {
     result should be(expected)
   }
 }
+
+class BundleSubSpec extends FlatSpec with Matchers {
+  "Bundle" should "only substitute body of expression" in {
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: "Bundle" should "substitute within the body of the bundle."

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala [hunk 4]
```diff
@@ -188,3 +188,38 @@ class EvalSubSpec extends FlatSpec with Matchers {
...
+    val source: Par = GPrivate()
+    val env         = Env.makeEnv(source)
+    val target      = Bundle(Send(ChanVar(BoundVar(0)), List(Par()), false, 0, BitSet(0)))
+    val result      = substitute(target)(env)
+    result should be(
+      Bundle(Send(Quote(source), List(Par()), false, 0, BitSet()))
+    )
+  }
+
+  it should "only substitute all Channels in body of express" in {
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: I think that should be expression.
it should "substitute all Channels inside body"

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala (hunk 2 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala (hunk 1 above)
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Fixed.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala (hunk 3 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/SubstituteTest.scala]: Fixed.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala (hunk 2 above)
Comment (goral09) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: Fixed error message. Currently for this specific case it throws:
```
coop.rchain.rholang.interpreter.NormalizerExceptions$UnexpectedBundleContent: Bundle's content shouldn't have free variables or wildcards.  Wildcards at positions: line: 0, column: 0. Free variables at positions: line: 0, column: 0.
```

I don't understand why there are only `0` -s though. Can you help? @guardbotmk3

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala [hunk 5]
```diff
@@ -165,21 +162,35 @@ object SpatialMatcher {
...
       }
     }
 
-  def listMatchItem[T](tlist: Seq[T],
-                       pattern: T,
-                       matcher: (T, T) => OptionalFreeMap[Unit]): NonDetFreeMap[Seq[T]] =
+  /** TODO: Consider moving this inside [[listMatchItem]]
+    */
+  private[this] def singleOut[A](vals: Seq[A]): Seq[(Seq[A], A, Seq[A])] =
+    vals.tails
```
[the remark this answers is not in the document]
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: I just moved this from some other place. I can't imagine having more than few dozen of elements in  it @guardbotmk3 ?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 6]
```diff
@@ -526,6 +530,27 @@ object ProcNormalizeMatcher {
         ProcVisitOutputs(input.par.prepend(foldedNew), bodyResult.knownFree)
       }
 
+      case b: PBundle =>
+        val targetResult = normalizeMatch(b.proc_, input.copy(par = VectorPar()))
+        if (targetResult.par.wildcard || targetResult.par.freeCount > 0) {
+          val errMsg = {
+            def at(variable: String, l: Int, col: Int): String =
+              s"line: $l, column: $col"
```
[the remark this answers is not in the document]
Comment (goral09) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Oh, nice spot. Thanks.

**Comment by goral09** [comment]
@pyrocto definitely in another PR. Addressed your comments.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: Well, the point of this function is to single out all the elements in O(n) time and space. I expect that what we'll want to do is if people complain, say that it's a known issue. Is there any chance of being on 2.12.5 for release?

If not, we can hand-roll the tails implementation if necessary.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 7]
```diff
@@ -526,6 +530,27 @@ object ProcNormalizeMatcher {
...
+              case (l, col) => at("", l, col)
+            }
+            val freeVarsPositions = targetResult.knownFree.env.map {
+              case (n, (_, _, line, col)) => at(s"`$n`", line, col)
+            }
+            wildcardsPositions.mkString(" Wildcards at positions: ", ", ", ".") ++
+              freeVarsPositions.mkString(" Free variables at positions: ", ", ", ".")
+          }
+          throw UnexpectedBundleContent(
+            s"Bundle's content shouldn't have free variables or wildcards.$errMsg")
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: So essentially we are relying on the normalizer to enforce the "bundling"? What if I bypass the normalizer and just send code straight to the interpreter or will that not be possible in our final design?

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: No, that won't be possible in the final design, but also:
The security of bundling doesn't come from there, it comes from the matcher, which won't match free variables inside of a bundle anyway.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala [hunk 8]
```diff
@@ -224,163 +237,182 @@ object SpatialMatcher {
...
-      for {
-        _           <- spatialMatch(target.chan.get, pattern.chan.get)
-        forcedYield <- foldMatch(target.data, pattern.data, (t: Par, p: Par) => spatialMatch(t, p))
-      } yield forcedYield
+  implicit val bundleSpatialMatcherInstance: SpatialMatcher[Bundle] = fromFunction[Bundle] {
+    (target, pattern) =>
+      if (pattern == target)
+        StateT.pure(Unit)
+      else {
+        StateT.liftF(None)
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala]: @guardbotmk3 Oh didn't see this line. Cool, good to know.

**Comment by goral09** [comment]
@KentShikama @guardbotmk3 @pyrocto Guys, I pushed one more commit (adding bundle support to the pretty printer) and now this PR needs another approval before merging.

**Comment by goral09** [comment]
Boom! Thanks @KentShikama :)

**Pull request merged**